README
$Id$
Last modified Sat Feb 21 00:44:29 2009 on violator
update count: 35

Hdf5io Python module 
====================
People familiar with Matlab(TM) and its load and save functions will find 
themselves at home with this small python module, I guess. It makes it 
possible to save your variables in "workspace" or from within a function 
to a hdf5 file, and later load them back in again.

Based on pyTables to be able to save and load data to/from hdf5 files.

Usage examples
==============

1. hdf5save
-----------
1a. To save all user create variables in current scope to a hdf5-file called 
"file.h5" type:

 hdf5save("file.h5")  

1b. To save specific variables in current scope to a hdf5-file called 
"file.h5" type:

 hdf5save("file.h5","var varstring")    

or:

 hdf5save("file.h5","var1","var2")   

1c. Names of variables to be saved can also be specified using wildcards '*':

 hdf5save("file.h5","v*")   

1d. Appending data to an existing hdf5 file is also possible. To invoke this
special mode add a + sign at the beginning of the file name:

 hdf5save("+file.h5","var")   

2. hdf5ls
-----------
To show the contents of a hdf5 file called "file.h5" type:

 hdf5ls("file.h5")

3. hdf5load
-----------
3a. To load the variables from a hdf5 file called "file.h5" type:

 hdf5load("file.h5")
 
3b. To load a specific variable from a hdf5 file called "file.h5" type:

 hdf5load("file.h5","var2")

or:

 hdf5load("file.h5","var1","var2")   

3c. Names of variables to be loaded can also be specified using wildcards '*':

 hdf5load("file.h5","v*")   

3d. To load variables in a specific group  from a hdf5 file called "file.h5" 
type:

 hdf5load("file.h5",None,"/group1")
                      

ChangeLog
2009-02-20 23:39  albert.thuswaldner

	* hdf5io/__init__.py, setup.py: Uped version

2009-02-20 23:38  albert.thuswaldner

	* release: fixed bug in release script

2009-02-20 23:24  albert.thuswaldner

	* hdf5io/hdf5io.py: added append option

2009-02-19 22:37  albert.thuswaldner

	* release: - made the release script more project independent

2009-02-08 22:26  albert.thuswaldner

	* release: updated release script to upload dist archive to
	  googlecode

2009-02-08 20:59  albert.thuswaldner

	* hdf5io/hdf5io.py: fixed hdf5save crash

2009-02-08 20:54  albert.thuswaldner

	* hdf5io/__init__.py, setup.py: Uped version

2009-02-05 22:08  albert.thuswaldner

	* hdf5io/__init__.py, setup.py: Uped version

2009-02-05 22:06  albert.thuswaldner

	* README: updated readme

2009-02-05 21:50  albert.thuswaldner

	* hdf5io/hdf5io.py: fixed wildcard feature

2009-02-04 12:13  albert.thuswaldner

	* hdf5io/hdf5io.py: fixed function documentation

2009-02-04 12:12  albert.thuswaldner

	* hdf5io/hdf5io.py: updated function documentation

2009-02-03 22:40  albert.thuswaldner

	* hdf5io/hdf5io.py: fixed some of the wildcard issue

2009-02-03 19:52  albert.thuswaldner

	* hdf5io/hdf5io.py: fixed condition in hdf5load

2009-02-02 23:25  albert.thuswaldner

	* hdf5io/hdf5io.py: commited changes to accept wildcards in
	  variable arg list

2009-01-29 23:34  albert.thuswaldner

	* release: completed release script

2009-01-29 08:47  albert.thuswaldner

	* release: included basic functionallity in release script

2009-01-29 08:29  albert.thuswaldner

	* release: property 'svn:keywords' and 'svn:executable' set on file

2009-01-29 08:29  albert.thuswaldner

	* release: property 'svn:keywords' set on file

2009-01-27 22:16  albert.thuswaldner

	* hdf5io/__init__.py: uped version number

2009-01-27 22:13  albert.thuswaldner

	* setup.py: uped version number

2009-01-27 22:09  albert.thuswaldner

	* hdf5io/hdf5io.py: Fixed load problems when loading from file with
	  nested groups

2009-01-14 21:33  albert.thuswaldner

	* TODO: updated todo

2009-01-14 21:32  albert.thuswaldner

	* hdf5io/hdf5io.py: updated help text

2009-01-13 23:56  albert.thuswaldner

	* TODO: updated todo

2009-01-13 23:49  albert.thuswaldner

	* TODO: updated todo

2009-01-13 23:40  albert.thuswaldner

	* hdf5io/hdf5io.py: - changed function call of hdf5load and
	  hdf5save to use *args
	  - buggfix: included create group functionality in hdf5save

2009-01-10 23:21  albert.thuswaldner

	* TODO: updated todo

2009-01-10 22:58  albert.thuswaldner

	* hdf5io/hdf5io.py: format long line in extractvars

2009-01-09 16:00  albert.thuswaldner

	* TODO: updated todo

2008-11-17 10:35  albert.thuswaldner

	* TODO: updated todo

2008-10-05 21:05  albert.thuswaldner

	* README: added example of loading variables from a specific group

2008-10-01 17:50  albert.thuswaldner

	* MANIFEST.in: dos2unix on MANIFEST.in

2008-10-01 17:40  albert.thuswaldner

	* README: minor fix on README

2008-10-01 17:35  albert.thuswaldner

	* hdf5io/__init__.py, test/hdf5test.py: fixed forgotten rename
	  hdf5info to hdf5ls

2008-09-27 17:14  albert.thuswaldner

	* README, TODO, test/hdf5test.py: changed name on hdf5info to
	  hdf5ls to be more inline with other implementations

2008-09-27 17:12  albert.thuswaldner

	* hdf5io/hdf5io.py: changed name on hdf5info to hdf5ls to be more
	  inline with other implementations

2008-08-20 20:57  albert.thuswaldner

	* hdf5io/__init__.py, setup.py: uped version 0.3

2008-08-20 20:53  albert.thuswaldner

	* TODO: checked done tasks

2008-08-20 20:50  albert.thuswaldner

	* hdf5io/hdf5io.py: - wrapped try block araound file open in info
	  and load functions

2008-08-19 22:03  albert.thuswaldner

	* hdf5io/hdf5io.py: - expanded help text for functions

2008-08-19 21:43  albert.thuswaldner

	* TODO: added todo task

2008-08-19 08:42  albert.thuswaldner

	* TODO: added todo task

2008-08-19 08:27  albert.thuswaldner

	* PKG-INFO, setup.py: updated package meta data

2008-08-18 20:18  albert.thuswaldner

	* PKG-INFO, hdf5io/__init__.py, setup.py: - uped version to 0.2

2008-08-18 20:11  albert.thuswaldner

	* README, TODO, hdf5io/hdf5io.py, test/hdf5test.py: - added option
	  to append variables to file
	  - added option to load speific varables from file

2008-08-18 18:31  albert.thuswaldner

	* hdf5io, setup.py, src: restructured repo

2008-08-18 18:25  albert.thuswaldner

	* INSTALL: editorial changes

2008-08-18 18:25  albert.thuswaldner

	* PKG-INFO, README: editorial changes

2008-08-18 07:31  albert.thuswaldner

	* TODO: - added new task to TODO

2008-08-18 07:14  albert.thuswaldner

	* MANIFEST.in: property 'svn:keywords' set on file

2008-08-18 06:43  albert.thuswaldner

	* setup.py: - fixed setup.py to work on windows

2008-08-17 20:28  albert.thuswaldner

	* src/__init__.py: - fixed init script

2008-08-17 19:54  albert.thuswaldner

	* INSTALL, PKG-INFO, README, setup.py, src/hdf5io.py: - edit on
	  description text

2008-08-17 19:17  albert.thuswaldner

	* __init__.py, setup.py, src/__init__.py: - fixed install

2008-08-17 18:45  albert.thuswaldner

	* COPYING, PKG-INFO, hdf5io.py, hdf5test.py, setup.py,
	  src/hdf5io.py, test/hdf5test.py: - restructured repo
	  - edit on setup.py
	  - added licence text and PKG-INFO

2008-08-17 18:42  albert.thuswaldner

	* src, test: added src/ and test/ dirs

2008-08-17 15:37  albert.thuswaldner

	* INSTALL, README, TODO, hdf5io.py, hdf5test.py, setup.py: -
	  updated descrption
	  - fixed hdf5save
	  - fixed test script

2008-08-17 15:01  albert.thuswaldner

	* INSTALL, README, TODO, __init__.py, hdf5io.py, hdf5test.py,
	  setup.py: set keyword properties on files

2008-08-17 14:59  albert.thuswaldner

	* INSTALL, README, TODO, __init__.py, hdf5io.py, hdf5test.py,
	  setup.py: initial commit

2008-08-17 14:52  

	* .: Initial directory structure.



test/hdf5test.py
#!/usr/bin/env python
# $Id$
# Last modified Wed Oct  1 19:33:55 2008 on violator
# update count: 11

from hdf5io import *

# test function
def hdf5test():
    # Create some test variables
    var = 1
    varstring='test string'
    vararray=[1, 32, 3]

    # Save to file
    hdf5save("test.h5")

    # Show file content 
    hdf5ls("test.h5")

    # Append a variable
    var2=67.3
    hdf5save("test.h5",'var2',"/",'a')

    # Show file content again 
    hdf5ls("test.h5")

    


hdf5io/hdf5io.py
#!/usr/bin/env python
# $Id$
# Last modified Sat Feb 21 00:45:02 2009 on violator
# update count: 471
#
# pyhdf5io - Python module containing hdf5 load and save functions.
# Copyright (C) 2008  Albert Thuswaldner
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
A set of I/O functions for the hdf5 file format.

Based on pyTables to be able to save and load data to/from hdf5 files.
"""
import inspect
import re
import tables

###############################################################################

def hdf5ls(filename):
    """
    Displays the contents of a hdf5 file.

    Syntax:
     hdf5ls('filename')
    """
    # Try to open and read from file 
    try:
        f=tables.openFile(filename,'r')

        try:
            for group in f.walkGroups():
                for node in f.listNodes(group):
                    print node
        finally:
          f.close()
    except IOError:
        print 'Cannot read:', filename

###############################################################################

def hdf5load(*args):
    """
    Loads variables from hdf5 file.
    
    Syntax:
     hdf5load('filename')
     hdf5load('filename', 'v*')
     hdf5load('filename', '/group')
     hdf5load('filename', 'var1', 'var2', ....)
     hdf5load('filename', '/group', 'var1', 'var2', ....)
     hdf5load('filename', '/group var1 var2 ....')
     hdf5load('filename', '/group var1', 'var2', 'var3 ....')

    Description:
     hdf5load loads variables from a hdf5 file to the namespace from where
     the function is called. The syntax is flexible which means that the
     function can be called in several different ways (see above). If only
     the file name is given, all variables are loaded. In addition, if
     a group name is supplied, all variables from that group is loaded.
     It is also possible to specify exactly which variables that should
     be loaded either by the complete variable name or by using wildcards '*'.
    """

    # Get dictonary from caller namespace
    dictvar=__magicLocals()

    # Extract input arguments
    inputargs=__extractargs(*args)
    filename=inputargs[0]
    groupname=inputargs[1]
    varnames=inputargs[2]

    # Try to open and read from file 
    try:
        f=tables.openFile(filename,'r')
        try:
            # Walk through group and create variables in workspace
            for group in f.walkGroups(groupname):
                # Walk through only the leaves (don't list the groups)
                for leaf in group._f_walkNodes('Leaf'):
                    if not varnames or varnames.match(leaf.name): 
                        dictvar[leaf.name] = leaf.read()
        finally:
           f.close()
    except IOError:
        print 'Cannot read:', filename

###############################################################################

def hdf5save(*args):
    """
    Saves variables to a hdf5 file.
    
    Syntax:
     hdf5save('filename')
     hdf5save('filename', 'v*')
     hdf5save('filename', '/group')
     hdf5save('filename', 'var1', 'var2', ....)
     hdf5save('filename', '/group', 'var1', 'var2', ....)
     hdf5save('filename', '/group var1 var2 ....')
     hdf5save('filename', '/group var1', 'var2', 'var3 ....')
     hdf5save('+filename')
     
    Description:
     hdf5save saves variables from the current namespace to a hdf5 file.
     The syntax is flexible which means that the function can be called
     in several different ways (see above). If only the file name is given,
     all variables are saved to the root group of the hdf5 file. In addition,
     if a group name is supplied, all variables can be saved to a specific
     group. It is also possible to specify exactly which variables that should
     be saved either by the complete variable name or by using wildcards '*'.
     Appending data to an existing hdf5 file is possible. To invoke this
     special mode just add a + sign to the beginning of the file name.  
    """
    # Get dictonary from caller namespace
    dictvar=__magicLocals()

    # Extract input arguments
    inputargs=__extractargs(*args)
    filename=inputargs[0]
    groupname=inputargs[1]
    varnames=inputargs[2]
    mode=inputargs[3]
    
    # Open file for writing
    f=tables.openFile(filename,mode)

    # Create group
    g=f.root  
    if groupname != "/":
        grouplist = groupname.split('/')
        for group in grouplist:
            # Check if group exists
            if g.__contains__(group): 
                g=g._f_getChild(group)
            else:
                if group:
                    g=f.createGroup(g,group)

    for key,value in dictvar.iteritems():
        if varnames.match(key) and __checkvars(key, value):
            f.createArray(g,key,value)

    # Close file
    f.close()

###############################################################################
# Helper functions

# Not used
def __magicGlobals(level=1):
    r"""Return the globals of the *caller*'s caller (default), or `level`
    callers up."""
    return inspect.getouterframes(inspect.currentframe())[1+level][0].f_globals

# used
def __magicLocals(level=1):
    r"""Return the locals of the *caller*'s caller (default) , or `level`
    callers up.
    """
    return inspect.getouterframes(inspect.currentframe())[1+level][0].f_locals

def __extractargs(*args):
    """check and identify input variables"""  

    groupname="/"
    varmatch='.'
    first=1
    mode="w"
    
    if len(args) >= 1:
        # transform args tuple to list
        arglist=list(args)
        # Pop the presumed filename from the args list 
        filename=arglist.pop(0)
        # Check if append (only applicable for hdf5save) 
        if filename[0] == "+":
            mode="a"
            filename=filename[1:]
            
        # loop the rest of the args list and extract group/variable names
        for arg in arglist:
            if type(arg) is str:
                varlist=arg.split()
                for var in varlist:
                    # Check if variable name list contains a group name
                    if var[0] == "/":
                        groupname=var
                    else:
                        if first:  
                            varmatch=var
                        else:
                            varmatch = varmatch+'|'+var
                            first=0
            else:
                raise ValueError, "variable input must be of type string"
    else:
        raise ValueError, "Too few arguments"

    # Compile regulare expression from match list
    if varmatch:
        varmatch='('+varmatch.replace("*",".*")+')'

    varnames=re.compile(varmatch)

    return (filename, groupname, varnames, mode)
        
def __checkvars(key, value):
    """check variables from global dictionary"""
    blacklist=['help','In','Out']
    if key[0] == "_" \
           or inspect.isclass(value) \
           or inspect.ismodule(value) \
           or inspect.isfunction(value) \
           or key in blacklist:
        return 0
    else:
        return 1

